presto-main/src/main/java/com/facebook/presto/sql/expressions/ExpressionOptimizerManager.java

```diff
@@ -0,0 +1,152 @@
+    {
+        try {
+            for (File file : listFiles(configurationDirectory)) {
+                if (file.isFile() && file.getName().endsWith(".properties")) {
+                    loadExpressionOptimizerFactory(file);
+                }
+            }
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException("Failed to load expression manager configuration", e);
+        }
+    }
+
+    private void loadExpressionOptimizerFactory(File configurationFile)
+            throws IOException
+    {
+        String name = getNameWithoutExtension(configurationFile.getName());
+        checkArgument(!isNullOrEmpty(name), "File name is empty, full path: %s", configurationFile.getAbsolutePath());
+        checkArgument(!name.equals(DEFAULT_OPTIMIZER_KEY), "Cannot name an expression optimizer instance default");
+
+        Map<String, String> properties = new HashMap<>(loadProperties(configurationFile));
+        String factoryName = properties.remove(EXPRESSION_MANAGER_FACTORY_NAME);
+        checkArgument(!isNullOrEmpty(factoryName), "%s does not contain %s", configurationFile, EXPRESSION_MANAGER_FACTORY_NAME);
+        checkArgument(expressionOptimizerFactories.containsKey(factoryName),
+                "ExpressionOptimizerFactory %s is not registered, registered factories: ", factoryName, expressionOptimizerFactories.keySet());
+        String defaultPropertyValue = properties.remove(DEFAULT_PROPERTY);
+        checkArgument(defaultPropertyValue == null || "true".equalsIgnoreCase(defaultPropertyValue) || "false".equalsIgnoreCase(defaultPropertyValue),
+                "Only true or false is allowed for %s, found: %s", DEFAULT_PROPERTY, defaultPropertyValue);
+        boolean isDefault = parseBoolean(defaultPropertyValue);
+
+        ExpressionOptimizer optimizer = expressionOptimizerFactories.get(factoryName).createOptimizer(
+                properties,
+                new ExpressionOptimizerContext(nodeManager, functionAndTypeManager, functionResolution));
+        expressionOptimizers.put(name, optimizer);
+        if (isDefault) {
+            checkArgument(defaultExpressionOptimizer.get().equals(DEFAULT_OPTIMIZER_KEY), "Default ExpressionOptimizer is already set, found: %s, attempted: %s", defaultExpressionOptimizer.get());
+            defaultExpressionOptimizer.set(name);
+        }
+    }
+
+    public void addExpressionOptimizerFactory(ExpressionOptimizerFactory expressionOptimizerFactory)
+    {
+        String name = expressionOptimizerFactory.getName();
+        checkArgument(
+                expressionOptimizerFactories.putIfAbsent(name, expressionOptimizerFactory) == null,
+                "ExpressionOptimizerFactory %s is already registered", name);
+    }
+
+    public ExpressionOptimizer getExpressionOptimizer(ConnectorSession connectorSession)
+    {
+        checkArgument(connectorSession instanceof FullConnectorSession, "connectorSession is not an instance of FullConnectorSession");
```

Comment:
i don't like assuming that a session is a connectorSession. it breaks the abstractions.  In this case all we care about is the session property. can we instead add a getSystemProperty() method to ConnectorSession?

Reply:
While I agree this is not ideal, I'm of the opinion that it's worse to expose `getSystemProperty` to the SPI, because this allows all plugins to access system properties that were previously only known to the engine or runtime.  While not ideal, I think this approach is a bit less worse because it's within `presto-main`, and we have examples of this being done for system connector and metadata tables provided to the JDBC driver.  I do think that we should probably think of a better abstraction for how we provide session properties to plugins which are not connectors, and I've added a TODO and will open a Github issue to track it.  Let me know if you think this is reasonable.

Reply:
Fair criticism and suggestion to punt on resolving this.  i do feel like the problem i'm talking about isn't about plugins that aren't connectors, it's easy enough to implement passing them ConnectorSessions. The thing that makes this require a ConnectorSession vs. a Session is that it gets called from connector optimizers.  and we had similar issue with stats stuff a while back.  The plugin separation break down when connectors call into stuff from presto-main.